### ISSUE TITLE
"Upload New Form Definition" dialog with the Dataset and a property changed in a form.

### ISSUE THREAD

**Issue by dbemke** [getodk/central#465]
#### Problem description
When a dataset and a property is changed in a form, the "Upload New Form Definition” dialog shows information about the changed property which might be misleading (no information about the changed dataset). Other properties weren’t changed but belong to the new Dataset so all of the properties are new for the Dataset.
#### URL of the page
https://staging.getodk.cloud/#/projects/20/forms/basic_entity
#### Steps to reproduce the problem
1. Upload and publish a form creating a Dataset (entity with properties). e.g.basic entity form version 202305180137898 in https://staging.getodk.cloud/#/projects/20/forms/basic_entity/versions
2. In the form add/remove a property and change the name of the Dataset (e.g. basic entity form version 20230518013 in https://staging.getodk.cloud/#/projects/20/forms/basic_entity/versions)
3. Upload the new version of a form and check the "Upload New Form Definition” dialog.
4. Click "Upload anyway”.
5. Check if the Dataset and properties are new in the draft section.
6. Publish the draft and check the dialog.
#### Screenshot
I checked it on staging and cloud server and there is a difference in the draft section.

Staging:
![uploadnewformdefinition](https://github.com/getodk/central/assets/56479916/f15a6d00-ab79-45ed-b1d9-00ea676f6875)
![draftsectionNewDatset](https://github.com/getodk/central/assets/56479916/a3cbccc0-9cd3-43bb-aadc-814ce7ba4ac2)
![publishDraft](https://github.com/getodk/central/assets/56479916/005fe285-1fae-47e9-9734-d588b36ac8c5)

Cloud (in the draft section properties aren't new- without +, not mentioned in publish draft dialog):
![cloudDraftSection](https://github.com/getodk/central/assets/56479916/912a9dfc-f7cc-4a41-bd44-9379330792f1)
![cloudPublishDraft](https://github.com/getodk/central/assets/56479916/8664c672-c77e-429c-92db-4b4f1d2fccd9)

#### Expected behavior
I’m not sure what should be the expected result (e.g. no dialog/ information about changing dataset) but the information about the changed property might be misleading as the changed Dataset is the major change in the form.
#### Central version shown in version.txt
https://test.getodk.cloud/ 
versions:
906addf66af230735e501ad73a1269012fdd7a5a (v2023.3.1)
+330132cf1ecb824b1fb88894287cb8edb5acefa5 client (v2023.3.0-2-g330132cf)
 ba6ee474f431b48e885fb035c9266792f06d0ad7 server (v2023.3.1)
https://staging.getodk.cloud/
versions:
versions:
f78da7a8b7a132ad2baa794a656da6ba8e40754a (v2023.3.1-6-gf78da7a)
+ad92f1148c2fd32d0faaac9c3571410a6c31094b client (v2023.3.0-13-gad92f114)
+0fb4ebd78afb8c31d8f16d8ad38183111ece264f server (v2023.3.1-31-g0fb4ebd7)
#### Browser version
Chrome 114.0.5735.198, Firefox 116.0.2

**Comment by matthew-white** [comment]
@ktuite, do you think this behavior changed with getodk/central-backend#933?

**Comment by sadiqkhoja** [comment]
The dialog doesn't show information about changes in Dataset or their properties (saveto).

The information shown in the dialog is correct because "thirdproperty" field was removed from the 2nd version

**Comment by ktuite** [comment]
> The information shown in the dialog is correct because "thirdproperty" field was removed from the 2nd version 

I think this is something we might want to talk about with the rest of team - if some previous version of a form populates a dataset property that later gets removed, that earlier version could still be out there accepting submissions and creating entities with that property, but there would be no indication in the Central UI that that form and that property were (historically) linked.

**Comment by matthew-white** [comment]
What stood out to me most of all is that the behavior seems to have changed: different things are shown on test as opposed to staging.

**Comment by ktuite** [comment]
Hmm, looking into the differences between staging and test. 

1) I'm not 100% sure what is up with the new green plusses missing from the properties when the dataset itself is supposedly new. This is only seen on test, right?  I think there may be a draft or deleted or otherwise unpublished form or something with those properties, which is why the dataset looks new but the properties don't.  I wasn't able to exactly recreate it but I did get into this other state (https://test.getodk.cloud/#/projects/445/forms/basic_entity_2/draft) where the dataset `basic2` is new and there are 5 out of 6 properties shown except there should only be 5 new ones.

Part of the diff response looks like:
```
    {
        "name": "third.property",
        "isNew": true,
        "inForm": false
    }
```


I think/hope this is fixed with the update I made to the `dataset-diff` endpoint. You can tell which is before/after the update by looking carefully at whether the property names are in alphabetical or not. 

2) I have an underwhelming technical explanation for why that one modal mentions deleted fields but not dataset properties. There are modals at two stages:

    a) when the form is uploaded. we check for deleted properties and DON'T actually save the form. They have to press the "upload anyway" button to re-POST it. we don't do dataset checks until the form draft is successfully uploaded. 

    b) the second modal is the publish confirmation modal. that should show the new dataset properties. 

I'm not sure if we can/should compute dataset info in time for that first upload modal. Well.. if we did do something, it would likely go in `Forms.checkStructuralChange`.

**Comment by ktuite** [comment]
Trying to clarify what I think this issue is about:

1. seeing a warning on form upload about a property being deleted, but no other dataset information available in the same modal

2. difference on staging vs. test between what is shown about dataset name and properties (getodk/central-backend#933 did change this, so as long as staging works, this should be ok)

**Comment by dbemke** [comment]
The steps in test.getodk.cloud were in https://test.getodk.cloud/#/projects/436 . I looked for an already created dataset, deleted, or draft form but it's not present in the project. One difference that I might have accidentally introduced is started testing changes not only in xlsx forms but also xml. So steps probably contained first xml , then xlsx with the new dataset form.

**Comment by matthew-white** [comment]
I followed the repro steps in staging after creating a new project: https://staging.getodk.cloud/#/projects/37/forms/basic_entity/draft. I thought I'd add a few thoughts.

> One difference that I might have accidentally introduced is that I started testing changes not only in xlsx forms but also xml.

XLSForms are converted to XML XForms anyway, so it shouldn't be a big deal. I took a look at the changes to the XML, and I didn't see anything surprising.

> When a dataset and a property is changed in a form, the "Upload New Form Definition” dialog shows information about the changed property which might be misleading (no information about the changed dataset).

When I uploaded the second version of the form, which deleted a form field and changed the entity list, I saw a warning about the field. I didn't see a warning that the form will no longer save to any of the properties of the first entity list. In other words, we warn when a field is removed, but not when an entity list or property is removed.

The reason why it's so important to mention the deletion of a field is that the field will no longer appear in the submission data by default. There is an effect on the structure of the form. People sometimes try to rename a field, but Central doesn't understand renaming and sees it as one field being deleted and another being added. It's partly because of this mismatch in the mental model that we show this warning.

Things are a little different with entity lists and properties. There's no way to delete an entity list or a property. If you try to rename an entity list or property, that will just create more things. You'll see what those changes would be when you go to publish the draft.

I do think there's some awkwardness here: the fact that some information is shown when you first upload the form, while other information is shown after you upload the form. I'm CCing @issa-tseng in case she has thoughts about ways we might improve that flow. That said, we haven't heard of users running into problems with renaming entity lists or properties. I'm sure we'll start hearing about such problems at some point, and users' feedback at that time may guide the changes we want to make.

> if some previous version of a form populates a dataset property that later gets removed, that earlier version could still be out there accepting submissions and creating entities with that property, but there would be no indication in the Central UI that that form and that property were (historically) linked.

I agree that there is potential for confusion here. @issa-tseng might have ideas about more information we could add to the UI about this. I also think we could wait to see how often and in what ways users encounter it: that might influence where we decide to show that information.

> I checked it on staging and cloud server and there is a difference in the draft section.

In my repro, I saw:

- Under "Entity List is updated by this Form:", the properties of `basic2` all have a plus next to them. That's different behavior from production, and I'm glad you realized that, @dbemke. I think this new behavior is also OK though.
- When I click "Publish Draft", I see a list of the properties of `basic2`. That's again different behavior from production, but I actually think it might be helpful to see the properties there.

> I'm not 100% sure what is up with the new green plusses missing from the properties when the dataset itself is supposedly new. This is only seen on test, right? I think there may be a draft or deleted or otherwise unpublished form or something with those properties, which is why the dataset looks new but the properties don't.

I think this is probably clear at this point, but just to be explicit, the plusses are expected to be missing in production. The test server is very similar to production currently, which is why we see them missing there. I don't think it's due to the state of the form.

> I did get into this other state (https://test.getodk.cloud/#/projects/445/forms/basic_entity_2/draft) where the dataset `basic2` is new and there are 5 out of 6 properties shown except there should only be 5 new ones.

That's exciting! I didn't see that in my repro. Did you first just rename `basic` to `basic2`, then change the properties in a second draft? I actually see a difference in test vs. staging when I do the following:

- Create a new project.
- Create a draft form that would create an entity list with some properties.
- Upload a new draft that would create the same entity list, but renames one of the properties.
- On test, I see the old name of the property in addition to the new, but on staging, I just see the new.

Do you think this was changed in getodk/central-backend#933? The behavior on staging seems like an improvement.

---

To summarize, I think the changes on staging are an improvement. Unless @issa-tseng can think of other improvements to make, I think we can close this issue. @dbemke, does that sound right? This discussion has been very useful: this change in behavior isn't something I had seen before.
